systemd/systemd.go

```diff
@@ -282,9 +282,21 @@ func (s *systemd) Stop(serviceName string, timeout time.Duration) error {
 	// and now wait for it to actually stop
 	stopped := false
 	max := time.Now().Add(timeout)
+
+	steps := stopSteps
+	toutSteps := int(timeout / stopDelay)
+	if toutSteps < steps {
+		steps = toutSteps
+	}
+	if steps < 1 {
+		steps = 1
+	}
+
 	for time.Now().Before(max) {
 		s.reporter.Notify(fmt.Sprintf("Waiting for %s to stop.", serviceName))
-		for i := 0; i < stopSteps; i++ {
+		for i := 0; i < steps; i++ {
```

Comment:
The logic above feels tricky to read. It took me a while to understand the bug you were fixing, and how it was supposed to work, and then whether it was actually working.

This kind of logic is always more clear if we think about it in terms of time rather than steps. I suggest replacing both of these loops by a single loop based on timers:
```
notify := time.NewTicker(notifyDelay)
check := time.NewTicker(checkDelay)
giveup := time.After(timeout)
for {
    select {
    case <-check:
            ...
    case <-notify:
            ...
    case <-giveup:
            return &Timeout{...}
    }
}
panic("unreachable")